// @switchboard-xyz/solana.js/lib/generated/oracle-program/instructions/aggregatorFunctionUpsert.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { TransactionInstruction } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export const layout = borsh.struct([
    types.AggregatorFunctionUpsertParams.layout("params"),
]);
export function aggregatorFunctionUpsert(program, args, accounts, programId = program.oracleProgramId) {
    const keys = [
        { pubkey: accounts.aggregator, isSigner: false, isWritable: true },
        { pubkey: accounts.function, isSigner: false, isWritable: false },
        {
            pubkey: accounts.functionEnclaveSigner,
            isSigner: true,
            isWritable: false,
        },
        { pubkey: accounts.verifier, isSigner: false, isWritable: false },
        { pubkey: accounts.payer, isSigner: true, isWritable: true },
        { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
    ];
    const identifier = Buffer.from([141, 189, 31, 87, 146, 172, 104, 58]);
    const buffer = Buffer.alloc(1000);
    const len = layout.encode({
        params: types.AggregatorFunctionUpsertParams.toEncodable(args.params),
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = new TransactionInstruction({ keys, programId, data });
    return ix;
}
