// @switchboard-xyz/solana.js/lib/generated/attestation-program/instructions/functionRequestInit.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { TransactionInstruction } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export const layout = borsh.struct([
    types.FunctionRequestInitParams.layout("params"),
]);
/** Request Actions */
export function functionRequestInit(program, args, accounts, programId = program.attestationProgramId) {
    const keys = [
        { pubkey: accounts.request, isSigner: true, isWritable: true },
        { pubkey: accounts.authority, isSigner: false, isWritable: false },
        { pubkey: accounts.function, isSigner: false, isWritable: true },
        { pubkey: accounts.functionAuthority, isSigner: true, isWritable: false },
        { pubkey: accounts.escrow, isSigner: false, isWritable: true },
        { pubkey: accounts.mint, isSigner: false, isWritable: false },
        { pubkey: accounts.state, isSigner: false, isWritable: false },
        { pubkey: accounts.attestationQueue, isSigner: false, isWritable: false },
        { pubkey: accounts.payer, isSigner: true, isWritable: true },
        { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
        { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
        {
            pubkey: accounts.associatedTokenProgram,
            isSigner: false,
            isWritable: false,
        },
    ];
    const identifier = Buffer.from([118, 8, 251, 119, 88, 174, 81, 239]);
    const buffer = Buffer.alloc(1000);
    const len = layout.encode({
        params: types.FunctionRequestInitParams.toEncodable(args.params),
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = new TransactionInstruction({ keys, programId, data });
    return ix;
}
