// @switchboard-xyz/solana.js/lib/generated/attestation-program/instructions/functionServiceVerify.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { TransactionInstruction } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export const layout = borsh.struct([
    types.FunctionServiceVerifyParams.layout("params"),
]);
export function functionServiceVerify(program, args, accounts, programId = program.attestationProgramId) {
    const keys = [
        { pubkey: accounts.service, isSigner: false, isWritable: true },
        {
            pubkey: accounts.functionEnclaveSigner,
            isSigner: false,
            isWritable: false,
        },
        { pubkey: accounts.escrowWallet, isSigner: false, isWritable: true },
        { pubkey: accounts.escrowTokenWallet, isSigner: false, isWritable: true },
        { pubkey: accounts.function, isSigner: false, isWritable: true },
        {
            pubkey: accounts.functionEscrowTokenWallet,
            isSigner: false,
            isWritable: true,
        },
        { pubkey: accounts.verifierQuote, isSigner: false, isWritable: false },
        {
            pubkey: accounts.verifierEnclaveSigner,
            isSigner: true,
            isWritable: false,
        },
        { pubkey: accounts.verifierPermission, isSigner: false, isWritable: false },
        { pubkey: accounts.attestationQueue, isSigner: false, isWritable: false },
        { pubkey: accounts.receiver, isSigner: false, isWritable: true },
        { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
    ];
    const identifier = Buffer.from([58, 127, 172, 41, 218, 178, 104, 226]);
    const buffer = Buffer.alloc(1000);
    const len = layout.encode({
        params: types.FunctionServiceVerifyParams.toEncodable(args.params),
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = new TransactionInstruction({ keys, programId, data });
    return ix;
}
