// @switchboard-xyz/solana.js/lib/generated/attestation-program/instructions/functionInit.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { TransactionInstruction } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export const layout = borsh.struct([types.FunctionInitParams.layout("params")]);
export function functionInit(program, args, accounts, programId = program.attestationProgramId) {
    const keys = [
        { pubkey: accounts.function, isSigner: false, isWritable: true },
        { pubkey: accounts.addressLookupTable, isSigner: false, isWritable: true },
        { pubkey: accounts.authority, isSigner: false, isWritable: false },
        { pubkey: accounts.attestationQueue, isSigner: false, isWritable: false },
        { pubkey: accounts.payer, isSigner: true, isWritable: true },
        { pubkey: accounts.escrowWallet, isSigner: false, isWritable: true },
        {
            pubkey: accounts.escrowWalletAuthority,
            isSigner: true,
            isWritable: false,
        },
        { pubkey: accounts.escrowTokenWallet, isSigner: false, isWritable: true },
        { pubkey: accounts.mint, isSigner: false, isWritable: false },
        { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
        {
            pubkey: accounts.associatedTokenProgram,
            isSigner: false,
            isWritable: false,
        },
        { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
        {
            pubkey: accounts.addressLookupProgram,
            isSigner: false,
            isWritable: false,
        },
    ];
    const identifier = Buffer.from([0, 20, 30, 24, 100, 146, 13, 162]);
    const buffer = Buffer.alloc(1000);
    const len = layout.encode({
        params: types.FunctionInitParams.toEncodable(args.params),
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = new TransactionInstruction({ keys, programId, data });
    return ix;
}
