// @switchboard-xyz/solana.js/lib/generated/attestation-program/instructions/serviceWorkerInit.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { TransactionInstruction } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export const layout = borsh.struct([
    types.ServiceWorkerInitParams.layout("params"),
]);
export function serviceWorkerInit(program, args, accounts, programId = program.attestationProgramId) {
    const keys = [
        { pubkey: accounts.serviceWorker, isSigner: true, isWritable: true },
        { pubkey: accounts.attestationQueue, isSigner: false, isWritable: false },
        { pubkey: accounts.rewardEscrowWallet, isSigner: false, isWritable: true },
        {
            pubkey: accounts.rewardEscrowWalletAuthority,
            isSigner: true,
            isWritable: false,
        },
        {
            pubkey: accounts.rewardEscrowTokenWallet,
            isSigner: false,
            isWritable: true,
        },
        { pubkey: accounts.mint, isSigner: false, isWritable: false },
        { pubkey: accounts.authority, isSigner: false, isWritable: false },
        { pubkey: accounts.payer, isSigner: true, isWritable: true },
        { pubkey: accounts.systemProgram, isSigner: false, isWritable: false },
        { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
        {
            pubkey: accounts.associatedTokenProgram,
            isSigner: false,
            isWritable: false,
        },
    ];
    const identifier = Buffer.from([224, 41, 202, 107, 100, 255, 213, 220]);
    const buffer = Buffer.alloc(1000);
    const len = layout.encode({
        params: types.ServiceWorkerInitParams.toEncodable(args.params),
    }, buffer);
    const data = Buffer.concat([identifier, buffer]).slice(0, 8 + len);
    const ix = new TransactionInstruction({ keys, programId, data });
    return ix;
}
