// @switchboard-xyz/solana.js/lib/accounts/attestationPermissionAccount.js
import * as errors from "../errors.js";
import * as types from "../generated/attestation-program/index.js";
import { TransactionObject } from "../TransactionObject.js";
import { Account } from "./account.js";
import { PublicKey, SystemProgram } from "@solana/web3.js";
/**
 *  Account type dictating the level of permissions between a granter and a grantee.
 *
 *  Data: {@linkcode types.AttestationPermissionAccountData}
 */
export class AttestationPermissionAccount extends Account {
    static accountName = "AttestationPermissionAccountData";
    /**
     *  Load an existing PermissionAccount with its current on-chain state
     */
    static async load(program, authority, granter, grantee) {
        const account = AttestationPermissionAccount.fromSeed(program, typeof authority === "string" ? new PublicKey(authority) : authority, typeof granter === "string" ? new PublicKey(granter) : granter, typeof grantee === "string" ? new PublicKey(grantee) : grantee);
        const state = await account.loadData();
        return [account, state];
    }
    /**
     *  Loads an AttestationPermissionAccount from the expected PDA seed format.
     *
     *  @param program The Switchboard program for the current connection.
     *  @param authority The authority pubkey to be incorporated into the account seed.
     *  @param granter The granter pubkey to be incorporated into the account seed.
     *  @param grantee The grantee pubkey to be incorporated into the account seed.
     *
     *  @return AttestationPermissionAccount and PDA bump.
     */
    static fromSeed(program, authority, granter, grantee) {
        const [publicKey, bump] = PublicKey.findProgramAddressSync([
            Buffer.from("PermissionAccountData"),
            authority.toBytes(),
            granter.toBytes(),
            grantee.toBytes(),
        ], program.attestationProgramId);
        return new AttestationPermissionAccount(program, publicKey);
    }
    static createInstruction(program, payer, params, options) {
        const authority = params.authority ?? payer;
        const account = AttestationPermissionAccount.fromSeed(program, authority, params.granter, params.grantee);
        const instruction = types.attestationPermissionInit(program, { params: {} }, {
            permission: account.publicKey,
            attestationQueue: params.granter,
            node: params.grantee,
            authority,
            payer,
            systemProgram: SystemProgram.programId,
        });
        return [account, new TransactionObject(payer, [instruction], [], options)];
    }
    static async create(program, params, options) {
        const [account, txnObject] = this.createInstruction(program, program.walletPubkey, params, options);
        const txSignature = await program.signAndSend(txnObject, options);
        return [account, txSignature];
    }
    /**
     *  Retrieve and decode the {@linkcode types.AttestationPermissionAccountData} stored in this account.
     */
    async loadData() {
        const data = await types.AttestationPermissionAccountData.fetch(this.program, this.publicKey);
        if (data)
            return data;
        throw new errors.AccountNotFoundError("Permissions (Attestation)", this.publicKey);
    }
    /**
     *  Produces the instruction to set the permission in the AttestationPermissionAccount
     */
    setInstruction(payer, params, options) {
        // const data = await this.loadData();
        return new TransactionObject(payer, [
            types.attestationPermissionSet(this.program, {
                params: {
                    permission: params.permission.discriminator,
                    enable: params.enable,
                },
            }, {
                permission: this.publicKey,
                authority: params.queueAuthority
                    ? params.queueAuthority.publicKey
                    : payer,
                attestationQueue: params.queue,
                grantee: params.enclave,
            }),
        ], params.queueAuthority ? [params.queueAuthority] : [], options);
    }
    /**
     *  Sets the permission in the AttestationPermissionAccount
     */
    async set(params, options) {
        const setTxn = await this.setInstruction(this.program.walletPubkey, params, options);
        const txnSignature = await this.program.signAndSend(setTxn, options);
        return txnSignature;
    }
}
