// @switchboard-xyz/solana.js/lib/accounts/functionServiceAccount.js
import * as errors from "../errors.js";
import * as types from "../generated/attestation-program/index.js";
import { NativeMint } from "../mint.js";
import { TransactionObject } from "../TransactionObject.js";
import { Account } from "./account.js";
import { SwitchboardWallet } from "./switchboardWallet.js";
import * as anchor from "@coral-xyz/anchor";
import { ASSOCIATED_TOKEN_PROGRAM_ID, TOKEN_PROGRAM_ID, } from "@solana/spl-token";
import { PublicKey } from "@solana/web3.js";
import { Keypair, SystemProgram } from "@solana/web3.js";
import { BN, parseRawMrEnclave } from "@switchboard-xyz/common";
/**
 * Account type representing a Switchboard Function.
 *
 * Data: {@linkcode types.FunctionServiceAccountData}
 */
export class FunctionServiceAccount extends Account {
    static accountName = "FunctionServiceAccountData";
    switchboardWallet = undefined;
    /**
     *  Retrieve and decode the {@linkcode types.FunctionServiceAccountData} stored in this account.
     */
    async loadData() {
        const data = await types.FunctionServiceAccountData.fetch(this.program, this.publicKey);
        if (!data) {
            throw new errors.AccountNotFoundError("Routine", this.publicKey);
        }
        // cache this so we dont need to always re-fetch
        this.switchboardWallet = new SwitchboardWallet(this.program, data.escrowWallet);
        return data;
    }
    async getSwitchboardWallet() {
        if (this.switchboardWallet) {
            return this.switchboardWallet;
        }
        const data = await this.loadData();
        if (this.switchboardWallet) {
            return this.switchboardWallet;
        }
        return new SwitchboardWallet(this.program, data.escrowWallet);
    }
    static async load(program, address) {
        const functionAccount = new FunctionServiceAccount(program, address);
        const state = await functionAccount.loadData();
        return [functionAccount, state];
    }
    static async createInstruction(program, payer, params, wallet, options) {
        // TODO: Calculate the max size of data we can support up front then split into multiple txns
        const authorityPubkey = params.authority ?? payer;
        const serviceKeypair = params.keypair ?? Keypair.generate();
        await program.verifyNewKeypair(serviceKeypair);
        // TODO: verify this math / units
        const enclaveSize = new BN(Number.parseInt(params.enclaveSize.toString()) * 1024);
        const cpu = new BN(Number.parseInt(params.cpu?.toString() ?? "1"));
        const functionState = await params.functionAccount.loadData();
        let escrowWallet;
        let escrowWalletAuthority;
        if (wallet) {
            escrowWallet = wallet;
            escrowWalletAuthority = (await escrowWallet.loadData()).authority;
            if (!escrowWalletAuthority.equals(payer) &&
                !escrowWalletAuthority.equals(authorityPubkey)) {
                throw new errors.IncorrectAuthority(escrowWalletAuthority, authorityPubkey);
            }
        }
        else {
            escrowWallet = SwitchboardWallet.fromSeed(program, functionState.attestationQueue, authorityPubkey, params.functionAccount.publicKey.toBytes());
            escrowWalletAuthority = authorityPubkey;
        }
        const instruction = types.functionServiceInit(program, {
            params: {
                // Metadata Config
                name: new Uint8Array(Buffer.from(params.name ?? "", "utf8")),
                metadata: new Uint8Array(Buffer.from(params.metadata ?? "", "utf8")),
                enclaveSize,
                cpu,
                maxContainerParamsLen: null,
                containerParams: new Uint8Array(params.containerParams ?? Buffer.from("")),
            },
        }, {
            service: serviceKeypair.publicKey,
            authority: authorityPubkey,
            function: params.functionAccount.publicKey,
            functionAuthority: functionState.authority,
            escrowWallet: escrowWallet.publicKey,
            escrowWalletAuthority: escrowWalletAuthority,
            escrowTokenWallet: anchor.utils.token.associatedAddress({
                mint: program.mint.address,
                owner: escrowWallet.publicKey,
            }),
            mint: program.mint.address,
            attestationQueue: functionState.attestationQueue,
            payer,
            systemProgram: SystemProgram.programId,
            tokenProgram: TOKEN_PROGRAM_ID,
            associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        });
        return [
            new FunctionServiceAccount(program, serviceKeypair.publicKey),
            new TransactionObject(payer, [instruction], [serviceKeypair], options),
        ];
    }
    static async create(program, params, wallet, options) {
        const [account, txnObject] = await this.createInstruction(program, program.walletPubkey, params, wallet, options);
        const txSignature = await program.signAndSend(txnObject, options);
        return [account, txSignature];
    }
    async getBalance() {
        const switchboardWallet = await this.getSwitchboardWallet();
        return switchboardWallet.getBalance();
    }
    async getBalanceBN() {
        const switchboardWallet = await this.getSwitchboardWallet();
        return switchboardWallet.getBalanceBN();
    }
    async setConfigInstruction(payer, params, options) {
        const serviceState = await this.loadData();
        let name = null;
        if (params.name !== undefined) {
            name = new Uint8Array(Buffer.from(params.name ?? "", "utf8"));
        }
        let metadata = null;
        if (params.metadata !== undefined) {
            metadata = new Uint8Array(Buffer.from(params.metadata ?? "", "utf8"));
        }
        let enclaveSize = null;
        if (params.enclaveSize !== undefined) {
            enclaveSize = new BN(Number.parseInt(params.enclaveSize.toString()) * 1024);
        }
        let cpu = null;
        if (params.cpu !== undefined) {
            cpu = new BN(Number.parseInt(params.cpu.toString()));
        }
        let containerParams = null;
        if (params.containerParams !== undefined) {
            containerParams = new Uint8Array(Buffer.from(params.containerParams ?? "", "utf8"));
        }
        const setConfigIxn = types.functionServiceSetConfig(this.program, {
            params: {
                name,
                metadata,
                enclaveSize,
                cpu,
                containerParams,
            },
        }, {
            service: this.publicKey,
            function: serviceState.function,
            serviceWorker: PublicKey.default.equals(serviceState.serviceWorker)
                ? this.program
                    .attestationProgramId /** Use programID for optional accounts */
                : serviceState.serviceWorker,
            authority: serviceState.authority,
        });
        return new TransactionObject(payer, [setConfigIxn], params?.authority ? [params.authority] : [], options);
    }
    async setConfig(params, options) {
        return await this.setConfigInstruction(this.program.walletPubkey, params, options).then((txn) => this.program.signAndSend(txn, options));
    }
    async fundInstruction(payer, params, options) {
        const switchboardWallet = await this.getSwitchboardWallet();
        return await switchboardWallet.fundInstruction(payer, { ...params, resources: [this.publicKey] }, options);
    }
    async fund(params, options) {
        return await this.fundInstruction(this.program.walletPubkey, params, options).then((txn) => this.program.signAndSend(txn, options));
    }
    async requestQuoteVerifyInstruction(payer, params, options) {
        const serviceState = await this.loadData();
        if (serviceState.authority.equals(PublicKey.default)) {
            throw new Error(`Services needs to be added to a ServiceWorker before it can be verified`);
        }
        if (params.authority &&
            serviceState.authority.equals(params.authority.publicKey)) {
            throw new errors.IncorrectAuthority(serviceState.authority, params.authority.publicKey);
        }
        const ixn = types.functionServiceRequestQuoteVerify(this.program, {
            params: {
                quoteRegistry: params.quoteRegistry
                    ? Buffer.from(params.quoteRegistry)
                    : null,
                registryKey: Buffer.from(params.registryKey),
            },
        }, {
            service: this.publicKey,
            serviceWorker: serviceState.serviceWorker,
            function: serviceState.function,
            attestationQueue: serviceState.attestationQueue,
            escrowWallet: serviceState.escrowWallet,
            escrowTokenWallet: serviceState.escrowTokenWallet,
            newEnclaveSigner: params.enclaveSigner.publicKey,
            authority: serviceState.authority,
        });
        return new TransactionObject(payer, [ixn], params.authority
            ? [params.enclaveSigner, params.authority]
            : [params.enclaveSigner], options);
    }
    async requestQuoteVerify(params, options) {
        return await this.requestQuoteVerifyInstruction(this.program.walletPubkey, params, options).then((txn) => this.program.signAndSend(txn, options));
    }
    verifyIxn(params) {
        const ixn = types.functionServiceVerify(this.program, {
            params: {
                observedTime: new BN(params.observedTime
                    ? params.observedTime
                    : Math.floor(Date.now() / 1000)),
                mrEnclave: Array.from(parseRawMrEnclave(params.mrEnclave)),
            },
        }, {
            service: this.publicKey,
            functionEnclaveSigner: params.functionEnclaveSigner,
            escrowWallet: params.escrowWallet,
            escrowTokenWallet: anchor.utils.token.associatedAddress({
                mint: NativeMint.address,
                owner: params.escrowWallet,
            }),
            function: params.function,
            functionEscrowTokenWallet: params.functionEscrow,
            verifierQuote: params.verifierQuote,
            verifierEnclaveSigner: params.verifierEnclaveSigner,
            verifierPermission: params.verifierPermission,
            attestationQueue: params.attestationQueue,
            receiver: params.receiver,
            tokenProgram: TOKEN_PROGRAM_ID,
        });
        return ixn;
    }
    async addWorkerInstruction(payer, authority, serviceWorker, options) {
        const serviceState = await this.loadData();
        const authorityPubkey = authority instanceof PublicKey ? authority : authority.publicKey;
        if (!authorityPubkey.equals(serviceState.authority)) {
            throw new errors.IncorrectAuthority(serviceState.authority, authorityPubkey);
        }
        const ixn = types.functionServiceAddWorker(this.program, { params: {} }, {
            service: this.publicKey,
            serviceWorker,
            function: serviceState.function,
            authority: authorityPubkey,
        });
        return new TransactionObject(payer, [ixn], authority instanceof PublicKey ? [] : [authority], options);
    }
    async addWorker(authority, serviceWorker, options) {
        return await this.addWorkerInstruction(this.program.walletPubkey, authority, serviceWorker, options).then((txn) => this.program.signAndSend(txn, options));
    }
    async removeWorkerInstruction(payer, authority, serviceWorker, options) {
        const serviceState = await this.loadData();
        const authorityPubkey = authority instanceof PublicKey ? authority : authority.publicKey;
        if (!authorityPubkey.equals(serviceState.authority)) {
            throw new errors.IncorrectAuthority(serviceState.authority, authorityPubkey);
        }
        const ixn = types.functionServiceRemoveWorker(this.program, { params: { idx: null } }, {
            service: this.publicKey,
            serviceWorker,
            function: serviceState.function,
            authority: authorityPubkey,
        });
        return new TransactionObject(payer, [ixn], authority instanceof PublicKey ? [] : [authority], options);
    }
    async removeWorker(authority, serviceWorker, options) {
        return await this.removeWorkerInstruction(this.program.walletPubkey, authority, serviceWorker, options).then((txn) => this.program.signAndSend(txn, options));
    }
}
