// @switchboard-xyz/solana.js/lib/accounts/functionRoutineAccount.js
import * as errors from "../errors.js";
import * as types from "../generated/attestation-program/index.js";
import { NativeMint } from "../mint.js";
import { TransactionObject } from "../TransactionObject.js";
import { parseRawBuffer } from "../utils.js";
import { Account } from "./account.js";
import { SwitchboardWallet } from "./switchboardWallet.js";
import * as anchor from "@coral-xyz/anchor";
import { ASSOCIATED_TOKEN_PROGRAM_ID, TOKEN_PROGRAM_ID, } from "@solana/spl-token";
import { Keypair, SystemProgram } from "@solana/web3.js";
import { BN, parseCronSchedule, parseRawMrEnclave, } from "@switchboard-xyz/common";
/**
 * Account type representing a Switchboard Function.
 *
 * Data: {@linkcode types.FunctionRoutineAccountData}
 */
export class FunctionRoutineAccount extends Account {
    static accountName = "FunctionRoutineAccountData";
    switchboardWallet = undefined;
    /**
     *  Retrieve and decode the {@linkcode types.FunctionRoutineAccountData} stored in this account.
     */
    async loadData() {
        const data = await types.FunctionRoutineAccountData.fetch(this.program, this.publicKey);
        if (!data) {
            throw new errors.AccountNotFoundError("Routine", this.publicKey);
        }
        // cache this so we dont need to always re-fetch
        this.switchboardWallet = new SwitchboardWallet(this.program, data.escrowWallet);
        return data;
    }
    async getSwitchboardWallet() {
        if (this.switchboardWallet) {
            return this.switchboardWallet;
        }
        const data = await this.loadData();
        if (this.switchboardWallet) {
            return this.switchboardWallet;
        }
        return new SwitchboardWallet(this.program, data.escrowWallet);
    }
    static async load(program, address) {
        const functionAccount = new FunctionRoutineAccount(program, address);
        const state = await functionAccount.loadData();
        return [functionAccount, state];
    }
    static async createInstruction(program, payer, params, wallet, options) {
        // TODO: Calculate the max size of data we can support up front then split into multiple txns
        const authorityPubkey = params.authority ?? payer;
        const routineKeypair = params.keypair ?? Keypair.generate();
        await program.verifyNewKeypair(routineKeypair);
        const functionState = await params.functionAccount.loadData();
        const cronSchedule = params.schedule
            ? Buffer.from(parseCronSchedule(params.schedule), "utf-8")
            : Buffer.from(Array(64).fill(0));
        let escrowWallet;
        let escrowWalletAuthority;
        if (wallet) {
            escrowWallet = wallet;
            escrowWalletAuthority = (await escrowWallet.loadData()).authority;
            if (!escrowWalletAuthority.equals(payer) &&
                !escrowWalletAuthority.equals(authorityPubkey)) {
                throw new errors.IncorrectAuthority(escrowWalletAuthority, authorityPubkey);
            }
        }
        else {
            escrowWallet = SwitchboardWallet.fromSeed(program, functionState.attestationQueue, authorityPubkey, params.functionAccount.publicKey.toBytes());
            escrowWalletAuthority = authorityPubkey;
        }
        const instruction = types.functionRoutineInit(program, {
            params: {
                // Metadata Config
                name: new Uint8Array(Buffer.from(params.name ?? "", "utf8")),
                metadata: new Uint8Array(Buffer.from(params.metadata ?? "", "utf8")),
                // Fees
                bounty: params?.bounty
                    ? typeof params.bounty === "number"
                        ? new BN(params.bounty)
                        : params.bounty
                    : null,
                // Execution Config
                schedule: cronSchedule,
                maxContainerParamsLen: null,
                containerParams: new Uint8Array(params.containerParams ?? Buffer.from("")),
            },
        }, {
            routine: routineKeypair.publicKey,
            authority: authorityPubkey,
            function: params.functionAccount.publicKey,
            functionAuthority: functionState.authority,
            escrowWallet: escrowWallet.publicKey,
            escrowWalletAuthority: escrowWalletAuthority,
            escrowTokenWallet: anchor.utils.token.associatedAddress({
                mint: program.mint.address,
                owner: escrowWallet.publicKey,
            }),
            mint: program.mint.address,
            attestationQueue: functionState.attestationQueue,
            payer,
            systemProgram: SystemProgram.programId,
            tokenProgram: TOKEN_PROGRAM_ID,
            associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        });
        return [
            new FunctionRoutineAccount(program, routineKeypair.publicKey),
            new TransactionObject(payer, [instruction], [routineKeypair], options),
        ];
    }
    static async create(program, params, wallet, options) {
        const [account, txnObject] = await this.createInstruction(program, program.walletPubkey, params, wallet, options);
        const txSignature = await program.signAndSend(txnObject, options);
        return [account, txSignature];
    }
    // public get wallet(): SwitchboardWallet {
    //   return new SwitchboardWallet(this.program, this.getEscrow());
    // }
    // public getEscrow(): PublicKey {
    //   return this.program.mint.getAssociatedAddress(this.publicKey);
    // }
    async getBalance() {
        const switchboardWallet = await this.getSwitchboardWallet();
        return switchboardWallet.getBalance();
    }
    async getBalanceBN() {
        const switchboardWallet = await this.getSwitchboardWallet();
        return switchboardWallet.getBalanceBN();
    }
    async setConfigInstruction(payer, params, options) {
        const routineState = await this.loadData();
        let name = null;
        if (params.name !== undefined) {
            name = new Uint8Array(Buffer.from(params.name ?? "", "utf8"));
        }
        let metadata = null;
        if (params.metadata !== undefined) {
            metadata = new Uint8Array(Buffer.from(params.metadata ?? "", "utf8"));
        }
        let containerParams = null;
        if (params.containerParams !== undefined) {
            containerParams = new Uint8Array(Buffer.from(params.containerParams ?? "", "utf8"));
        }
        const setConfigIxn = types.functionRoutineSetConfig(this.program, {
            params: {
                name,
                metadata,
                // Fees
                bounty: params?.bounty
                    ? typeof params.bounty === "number"
                        ? new BN(params.bounty)
                        : params.bounty
                    : null,
                schedule: params?.schedule
                    ? Buffer.from(parseCronSchedule(params.schedule), "utf-8")
                    : null,
                containerParams,
                appendContainerParams: params.appendContainerParams ?? false,
            },
        }, {
            routine: this.publicKey,
            authority: routineState.authority,
        });
        return new TransactionObject(payer, [setConfigIxn], params?.authority ? [params.authority] : [], options);
    }
    async setConfig(params, options) {
        return await this.setConfigInstruction(this.program.walletPubkey, params, options).then((txn) => this.program.signAndSend(txn, options));
    }
    async fundInstruction(payer, params, options) {
        const switchboardWallet = await this.getSwitchboardWallet();
        return await switchboardWallet.fundInstruction(payer, { ...params, resources: [this.publicKey] }, options);
    }
    async fund(params, options) {
        return await this.fundInstruction(this.program.walletPubkey, params, options).then((txn) => this.program.signAndSend(txn, options));
    }
    async withdrawInstruction(payer, amount, destinationWallet, options) {
        const switchboardWallet = await this.getSwitchboardWallet();
        const txn = await switchboardWallet.withdrawInstruction(payer, amount, destinationWallet, options);
        return txn;
    }
    async withdraw(amount, destinationWallet, options) {
        return await this.withdrawInstruction(this.program.walletPubkey, amount, destinationWallet, options).then((txn) => this.program.signAndSend(txn, options));
    }
    verifyIxn(params) {
        const ixn = types.functionRoutineVerify(this.program, {
            params: {
                observedTime: new BN(params.observedTime
                    ? params.observedTime
                    : Math.floor(Date.now() / 1000)),
                nextAllowedTimestamp: new BN(params.nextAllowedTimestamp),
                errorCode: params.isFailure ? 1 : 0,
                mrEnclave: Array.from(parseRawMrEnclave(params.mrEnclave)),
                containerParamsHash: Array.from(parseRawBuffer(params.containerParamsHash)),
            },
        }, {
            routine: this.publicKey,
            functionEnclaveSigner: params.functionEnclaveSigner,
            escrowWallet: params.escrowWallet,
            escrowTokenWallet: anchor.utils.token.associatedAddress({
                mint: NativeMint.address,
                owner: params.escrowWallet,
            }),
            function: params.function,
            functionEscrowTokenWallet: params.functionEscrow,
            verifierQuote: params.verifierQuote,
            verifierEnclaveSigner: params.verifierEnclaveSigner,
            verifierPermission: params.verifierPermission,
            attestationQueue: params.attestationQueue,
            receiver: params.receiver,
            tokenProgram: TOKEN_PROGRAM_ID,
        });
        return ixn;
    }
}
