// @switchboard-xyz/solana.js/lib/generated/oracle-program/accounts/RealmSpawnRecordAccountData.js
import * as borsh from "@coral-xyz/borsh"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class RealmSpawnRecordAccountData {
    ebuf;
    static discriminator = Buffer.from([
        229, 116, 99, 2, 145, 96, 5, 95,
    ]);
    static layout = borsh.struct([borsh.array(borsh.u8(), 256, "ebuf")]);
    constructor(fields) {
        this.ebuf = fields.ebuf;
    }
    static async fetch(program, address, programId = program.oracleProgramId) {
        const info = await program.connection.getAccountInfo(address);
        if (info === null) {
            return null;
        }
        if (!info.owner.equals(programId)) {
            throw new Error("account doesn't belong to this program");
        }
        return this.decode(info.data);
    }
    static async fetchMultiple(program, addresses, programId = program.oracleProgramId) {
        const infos = await program.connection.getMultipleAccountsInfo(addresses);
        return infos.map((info) => {
            if (info === null) {
                return null;
            }
            if (!info.owner.equals(programId)) {
                throw new Error("account doesn't belong to this program");
            }
            return this.decode(info.data);
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(RealmSpawnRecordAccountData.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = RealmSpawnRecordAccountData.layout.decode(data.slice(8));
        return new RealmSpawnRecordAccountData({
            ebuf: dec.ebuf,
        });
    }
    toJSON() {
        return {
            ebuf: this.ebuf,
        };
    }
    static fromJSON(obj) {
        return new RealmSpawnRecordAccountData({
            ebuf: obj.ebuf,
        });
    }
}
