// @switchboard-xyz/solana.js/lib/generated/oracle-program/accounts/SbState.js
import * as borsh from "@coral-xyz/borsh"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { PublicKey } from "@solana/web3.js";
export class SbState {
    /** The account authority permitted to make account changes. */
    authority;
    /** The token mint used for oracle rewards, aggregator leases, and other reward incentives. */
    tokenMint;
    /** Token vault used by the program to receive kickbacks. */
    tokenVault;
    /** The token mint used by the DAO. */
    daoMint;
    /** The PDA bump to derive the pubkey. */
    bump;
    /** Permitted enclave measurements */
    mrEnclaves;
    /** Reserved for future info. */
    ebuf;
    static discriminator = Buffer.from([
        159, 42, 192, 191, 139, 62, 168, 28,
    ]);
    static layout = borsh.struct([
        borsh.publicKey("authority"),
        borsh.publicKey("tokenMint"),
        borsh.publicKey("tokenVault"),
        borsh.publicKey("daoMint"),
        borsh.u8("bump"),
        borsh.array(borsh.array(borsh.u8(), 32), 6, "mrEnclaves"),
        borsh.array(borsh.u8(), 799, "ebuf"),
    ]);
    constructor(fields) {
        this.authority = fields.authority;
        this.tokenMint = fields.tokenMint;
        this.tokenVault = fields.tokenVault;
        this.daoMint = fields.daoMint;
        this.bump = fields.bump;
        this.mrEnclaves = fields.mrEnclaves;
        this.ebuf = fields.ebuf;
    }
    static async fetch(program, address, programId = program.oracleProgramId) {
        const info = await program.connection.getAccountInfo(address);
        if (info === null) {
            return null;
        }
        if (!info.owner.equals(programId)) {
            throw new Error("account doesn't belong to this program");
        }
        return this.decode(info.data);
    }
    static async fetchMultiple(program, addresses, programId = program.oracleProgramId) {
        const infos = await program.connection.getMultipleAccountsInfo(addresses);
        return infos.map((info) => {
            if (info === null) {
                return null;
            }
            if (!info.owner.equals(programId)) {
                throw new Error("account doesn't belong to this program");
            }
            return this.decode(info.data);
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(SbState.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = SbState.layout.decode(data.slice(8));
        return new SbState({
            authority: dec.authority,
            tokenMint: dec.tokenMint,
            tokenVault: dec.tokenVault,
            daoMint: dec.daoMint,
            bump: dec.bump,
            mrEnclaves: dec.mrEnclaves,
            ebuf: dec.ebuf,
        });
    }
    toJSON() {
        return {
            authority: this.authority.toString(),
            tokenMint: this.tokenMint.toString(),
            tokenVault: this.tokenVault.toString(),
            daoMint: this.daoMint.toString(),
            bump: this.bump,
            mrEnclaves: this.mrEnclaves,
            ebuf: this.ebuf,
        };
    }
    static fromJSON(obj) {
        return new SbState({
            authority: new PublicKey(obj.authority),
            tokenMint: new PublicKey(obj.tokenMint),
            tokenVault: new PublicKey(obj.tokenVault),
            daoMint: new PublicKey(obj.daoMint),
            bump: obj.bump,
            mrEnclaves: obj.mrEnclaves,
            ebuf: obj.ebuf,
        });
    }
}
