// @switchboard-xyz/solana.js/lib/generated/oracle-program/accounts/TaskSpecRecord.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class TaskSpecRecord {
    hash;
    static discriminator = Buffer.from([
        202, 10, 194, 236, 111, 47, 234, 48,
    ]);
    static layout = borsh.struct([types.Hash.layout("hash")]);
    constructor(fields) {
        this.hash = new types.Hash({ ...fields.hash });
    }
    static async fetch(program, address, programId = program.oracleProgramId) {
        const info = await program.connection.getAccountInfo(address);
        if (info === null) {
            return null;
        }
        if (!info.owner.equals(programId)) {
            throw new Error("account doesn't belong to this program");
        }
        return this.decode(info.data);
    }
    static async fetchMultiple(program, addresses, programId = program.oracleProgramId) {
        const infos = await program.connection.getMultipleAccountsInfo(addresses);
        return infos.map((info) => {
            if (info === null) {
                return null;
            }
            if (!info.owner.equals(programId)) {
                throw new Error("account doesn't belong to this program");
            }
            return this.decode(info.data);
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(TaskSpecRecord.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = TaskSpecRecord.layout.decode(data.slice(8));
        return new TaskSpecRecord({
            hash: types.Hash.fromDecoded(dec.hash),
        });
    }
    toJSON() {
        return {
            hash: this.hash.toJSON(),
        };
    }
    static fromJSON(obj) {
        return new TaskSpecRecord({
            hash: types.Hash.fromJSON(obj.hash),
        });
    }
}
