// @switchboard-xyz/solana.js/lib/generated/oracle-program/accounts/TaskSpecRecord.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskSpecRecord = void 0;
const types = __importStar(require("../types/index.cjs")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh")); // eslint-disable-line @typescript-eslint/no-unused-vars
class TaskSpecRecord {
    hash;
    static discriminator = Buffer.from([
        202, 10, 194, 236, 111, 47, 234, 48,
    ]);
    static layout = borsh.struct([types.Hash.layout("hash")]);
    constructor(fields) {
        this.hash = new types.Hash({ ...fields.hash });
    }
    static async fetch(program, address, programId = program.oracleProgramId) {
        const info = await program.connection.getAccountInfo(address);
        if (info === null) {
            return null;
        }
        if (!info.owner.equals(programId)) {
            throw new Error("account doesn't belong to this program");
        }
        return this.decode(info.data);
    }
    static async fetchMultiple(program, addresses, programId = program.oracleProgramId) {
        const infos = await program.connection.getMultipleAccountsInfo(addresses);
        return infos.map((info) => {
            if (info === null) {
                return null;
            }
            if (!info.owner.equals(programId)) {
                throw new Error("account doesn't belong to this program");
            }
            return this.decode(info.data);
        });
    }
    static decode(data) {
        if (!data.slice(0, 8).equals(TaskSpecRecord.discriminator)) {
            throw new Error("invalid account discriminator");
        }
        const dec = TaskSpecRecord.layout.decode(data.slice(8));
        return new TaskSpecRecord({
            hash: types.Hash.fromDecoded(dec.hash),
        });
    }
    toJSON() {
        return {
            hash: this.hash.toJSON(),
        };
    }
    static fromJSON(obj) {
        return new TaskSpecRecord({
            hash: types.Hash.fromJSON(obj.hash),
        });
    }
}
exports.TaskSpecRecord = TaskSpecRecord;
