// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/AttestationPermissionSetParams.js
import * as borsh from "@coral-xyz/borsh";
export class AttestationPermissionSetParams {
    permission;
    enable;
    constructor(fields) {
        this.permission = fields.permission;
        this.enable = fields.enable;
    }
    static layout(property) {
        return borsh.struct([borsh.u32("permission"), borsh.bool("enable")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new AttestationPermissionSetParams({
            permission: obj.permission,
            enable: obj.enable,
        });
    }
    static toEncodable(fields) {
        return {
            permission: fields.permission,
            enable: fields.enable,
        };
    }
    toJSON() {
        return {
            permission: this.permission,
            enable: this.enable,
        };
    }
    static fromJSON(obj) {
        return new AttestationPermissionSetParams({
            permission: obj.permission,
            enable: obj.enable,
        });
    }
    toEncodable() {
        return AttestationPermissionSetParams.toEncodable(this);
    }
}
