// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/AttestationPermissionSetParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AttestationPermissionSetParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
class AttestationPermissionSetParams {
    permission;
    enable;
    constructor(fields) {
        this.permission = fields.permission;
        this.enable = fields.enable;
    }
    static layout(property) {
        return borsh.struct([borsh.u32("permission"), borsh.bool("enable")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new AttestationPermissionSetParams({
            permission: obj.permission,
            enable: obj.enable,
        });
    }
    static toEncodable(fields) {
        return {
            permission: fields.permission,
            enable: fields.enable,
        };
    }
    toJSON() {
        return {
            permission: this.permission,
            enable: this.enable,
        };
    }
    static fromJSON(obj) {
        return new AttestationPermissionSetParams({
            permission: obj.permission,
            enable: obj.enable,
        });
    }
    toEncodable() {
        return AttestationPermissionSetParams.toEncodable(this);
    }
}
exports.AttestationPermissionSetParams = AttestationPermissionSetParams;
