// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/BoolWithLock.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = exports.fromJSON = exports.fromDecoded = exports.TrueLocked = exports.FalseLocked = exports.TrueUnlocked = exports.FalseUnlocked = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
class FalseUnlocked {
    static discriminator = 0;
    static kind = "FalseUnlocked";
    discriminator = 0;
    kind = "FalseUnlocked";
    toJSON() {
        return {
            kind: "FalseUnlocked",
        };
    }
    toEncodable() {
        return {
            FalseUnlocked: {},
        };
    }
}
exports.FalseUnlocked = FalseUnlocked;
class TrueUnlocked {
    static discriminator = 1;
    static kind = "TrueUnlocked";
    discriminator = 1;
    kind = "TrueUnlocked";
    toJSON() {
        return {
            kind: "TrueUnlocked",
        };
    }
    toEncodable() {
        return {
            TrueUnlocked: {},
        };
    }
}
exports.TrueUnlocked = TrueUnlocked;
class FalseLocked {
    static discriminator = 2;
    static kind = "FalseLocked";
    discriminator = 2;
    kind = "FalseLocked";
    toJSON() {
        return {
            kind: "FalseLocked",
        };
    }
    toEncodable() {
        return {
            FalseLocked: {},
        };
    }
}
exports.FalseLocked = FalseLocked;
class TrueLocked {
    static discriminator = 3;
    static kind = "TrueLocked";
    discriminator = 3;
    kind = "TrueLocked";
    toJSON() {
        return {
            kind: "TrueLocked",
        };
    }
    toEncodable() {
        return {
            TrueLocked: {},
        };
    }
}
exports.TrueLocked = TrueLocked;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("FalseUnlocked" in obj) {
        return new FalseUnlocked();
    }
    if ("TrueUnlocked" in obj) {
        return new TrueUnlocked();
    }
    if ("FalseLocked" in obj) {
        return new FalseLocked();
    }
    if ("TrueLocked" in obj) {
        return new TrueLocked();
    }
    throw new Error("Invalid enum object");
}
exports.fromDecoded = fromDecoded;
function fromJSON(obj) {
    switch (obj.kind) {
        case "FalseUnlocked": {
            return new FalseUnlocked();
        }
        case "TrueUnlocked": {
            return new TrueUnlocked();
        }
        case "FalseLocked": {
            return new FalseLocked();
        }
        case "TrueLocked": {
            return new TrueLocked();
        }
    }
}
exports.fromJSON = fromJSON;
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "FalseUnlocked"),
        borsh.struct([], "TrueUnlocked"),
        borsh.struct([], "FalseLocked"),
        borsh.struct([], "TrueLocked"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
exports.layout = layout;
