// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/EnclaveRotationStatus.js
import * as borsh from "@coral-xyz/borsh";
export class None {
    static discriminator = 0;
    static kind = "None";
    discriminator = 0;
    kind = "None";
    toJSON() {
        return {
            kind: "None",
        };
    }
    toEncodable() {
        return {
            None: {},
        };
    }
}
export class InProgress {
    static discriminator = 1;
    static kind = "InProgress";
    discriminator = 1;
    kind = "InProgress";
    toJSON() {
        return {
            kind: "InProgress",
        };
    }
    toEncodable() {
        return {
            InProgress: {},
        };
    }
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("None" in obj) {
        return new None();
    }
    if ("InProgress" in obj) {
        return new InProgress();
    }
    throw new Error("Invalid enum object");
}
export function fromJSON(obj) {
    switch (obj.kind) {
        case "None": {
            return new None();
        }
        case "InProgress": {
            return new InProgress();
        }
    }
}
export function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "None"),
        borsh.struct([], "InProgress"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
