// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionServiceSetConfigParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionServiceSetConfigParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class FunctionServiceSetConfigParams {
    name;
    metadata;
    enclaveSize;
    cpu;
    containerParams;
    constructor(fields) {
        this.name = fields.name;
        this.metadata = fields.metadata;
        this.enclaveSize = fields.enclaveSize;
        this.cpu = fields.cpu;
        this.containerParams = fields.containerParams;
    }
    static layout(property) {
        return borsh.struct([
            borsh.option(borsh.vecU8(), "name"),
            borsh.option(borsh.vecU8(), "metadata"),
            borsh.option(borsh.u64(), "enclaveSize"),
            borsh.option(borsh.u64(), "cpu"),
            borsh.option(borsh.vecU8(), "containerParams"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionServiceSetConfigParams({
            name: (obj.name &&
                new Uint8Array(obj.name.buffer, obj.name.byteOffset, obj.name.length)) ||
                null,
            metadata: (obj.metadata &&
                new Uint8Array(obj.metadata.buffer, obj.metadata.byteOffset, obj.metadata.length)) ||
                null,
            enclaveSize: obj.enclaveSize,
            cpu: obj.cpu,
            containerParams: (obj.containerParams &&
                new Uint8Array(obj.containerParams.buffer, obj.containerParams.byteOffset, obj.containerParams.length)) ||
                null,
        });
    }
    static toEncodable(fields) {
        return {
            name: (fields.name &&
                Buffer.from(fields.name.buffer, fields.name.byteOffset, fields.name.length)) ||
                null,
            metadata: (fields.metadata &&
                Buffer.from(fields.metadata.buffer, fields.metadata.byteOffset, fields.metadata.length)) ||
                null,
            enclaveSize: fields.enclaveSize,
            cpu: fields.cpu,
            containerParams: (fields.containerParams &&
                Buffer.from(fields.containerParams.buffer, fields.containerParams.byteOffset, fields.containerParams.length)) ||
                null,
        };
    }
    toJSON() {
        return {
            name: (this.name && Array.from(this.name.values())) || null,
            metadata: (this.metadata && Array.from(this.metadata.values())) || null,
            enclaveSize: (this.enclaveSize && this.enclaveSize.toString()) || null,
            cpu: (this.cpu && this.cpu.toString()) || null,
            containerParams: (this.containerParams && Array.from(this.containerParams.values())) ||
                null,
        };
    }
    static fromJSON(obj) {
        return new FunctionServiceSetConfigParams({
            name: (obj.name && Uint8Array.from(obj.name)) || null,
            metadata: (obj.metadata && Uint8Array.from(obj.metadata)) || null,
            enclaveSize: (obj.enclaveSize && new common_1.BN(obj.enclaveSize)) || null,
            cpu: (obj.cpu && new common_1.BN(obj.cpu)) || null,
            containerParams: (obj.containerParams && Uint8Array.from(obj.containerParams)) || null,
        });
    }
    toEncodable() {
        return FunctionServiceSetConfigParams.toEncodable(this);
    }
}
exports.FunctionServiceSetConfigParams = FunctionServiceSetConfigParams;
