// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionRequestInitAndTriggerParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionRequestInitAndTriggerParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class FunctionRequestInitAndTriggerParams {
    bounty;
    slotsUntilExpiration;
    maxContainerParamsLen;
    containerParams;
    garbageCollectionSlot;
    validAfterSlot;
    constructor(fields) {
        this.bounty = fields.bounty;
        this.slotsUntilExpiration = fields.slotsUntilExpiration;
        this.maxContainerParamsLen = fields.maxContainerParamsLen;
        this.containerParams = fields.containerParams;
        this.garbageCollectionSlot = fields.garbageCollectionSlot;
        this.validAfterSlot = fields.validAfterSlot;
    }
    static layout(property) {
        return borsh.struct([
            borsh.option(borsh.u64(), "bounty"),
            borsh.option(borsh.u64(), "slotsUntilExpiration"),
            borsh.option(borsh.u32(), "maxContainerParamsLen"),
            borsh.option(borsh.vecU8(), "containerParams"),
            borsh.option(borsh.u64(), "garbageCollectionSlot"),
            borsh.option(borsh.u64(), "validAfterSlot"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionRequestInitAndTriggerParams({
            bounty: obj.bounty,
            slotsUntilExpiration: obj.slotsUntilExpiration,
            maxContainerParamsLen: obj.maxContainerParamsLen,
            containerParams: (obj.containerParams &&
                new Uint8Array(obj.containerParams.buffer, obj.containerParams.byteOffset, obj.containerParams.length)) ||
                null,
            garbageCollectionSlot: obj.garbageCollectionSlot,
            validAfterSlot: obj.validAfterSlot,
        });
    }
    static toEncodable(fields) {
        return {
            bounty: fields.bounty,
            slotsUntilExpiration: fields.slotsUntilExpiration,
            maxContainerParamsLen: fields.maxContainerParamsLen,
            containerParams: (fields.containerParams &&
                Buffer.from(fields.containerParams.buffer, fields.containerParams.byteOffset, fields.containerParams.length)) ||
                null,
            garbageCollectionSlot: fields.garbageCollectionSlot,
            validAfterSlot: fields.validAfterSlot,
        };
    }
    toJSON() {
        return {
            bounty: (this.bounty && this.bounty.toString()) || null,
            slotsUntilExpiration: (this.slotsUntilExpiration && this.slotsUntilExpiration.toString()) ||
                null,
            maxContainerParamsLen: this.maxContainerParamsLen,
            containerParams: (this.containerParams && Array.from(this.containerParams.values())) ||
                null,
            garbageCollectionSlot: (this.garbageCollectionSlot && this.garbageCollectionSlot.toString()) ||
                null,
            validAfterSlot: (this.validAfterSlot && this.validAfterSlot.toString()) || null,
        };
    }
    static fromJSON(obj) {
        return new FunctionRequestInitAndTriggerParams({
            bounty: (obj.bounty && new common_1.BN(obj.bounty)) || null,
            slotsUntilExpiration: (obj.slotsUntilExpiration && new common_1.BN(obj.slotsUntilExpiration)) || null,
            maxContainerParamsLen: obj.maxContainerParamsLen,
            containerParams: (obj.containerParams && Uint8Array.from(obj.containerParams)) || null,
            garbageCollectionSlot: (obj.garbageCollectionSlot && new common_1.BN(obj.garbageCollectionSlot)) ||
                null,
            validAfterSlot: (obj.validAfterSlot && new common_1.BN(obj.validAfterSlot)) || null,
        });
    }
    toEncodable() {
        return FunctionRequestInitAndTriggerParams.toEncodable(this);
    }
}
exports.FunctionRequestInitAndTriggerParams = FunctionRequestInitAndTriggerParams;
