// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionRequestTriggerParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionRequestTriggerParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class FunctionRequestTriggerParams {
    bounty;
    slotsUntilExpiration;
    validAfterSlot;
    constructor(fields) {
        this.bounty = fields.bounty;
        this.slotsUntilExpiration = fields.slotsUntilExpiration;
        this.validAfterSlot = fields.validAfterSlot;
    }
    static layout(property) {
        return borsh.struct([
            borsh.option(borsh.u64(), "bounty"),
            borsh.option(borsh.u64(), "slotsUntilExpiration"),
            borsh.option(borsh.u64(), "validAfterSlot"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionRequestTriggerParams({
            bounty: obj.bounty,
            slotsUntilExpiration: obj.slotsUntilExpiration,
            validAfterSlot: obj.validAfterSlot,
        });
    }
    static toEncodable(fields) {
        return {
            bounty: fields.bounty,
            slotsUntilExpiration: fields.slotsUntilExpiration,
            validAfterSlot: fields.validAfterSlot,
        };
    }
    toJSON() {
        return {
            bounty: (this.bounty && this.bounty.toString()) || null,
            slotsUntilExpiration: (this.slotsUntilExpiration && this.slotsUntilExpiration.toString()) ||
                null,
            validAfterSlot: (this.validAfterSlot && this.validAfterSlot.toString()) || null,
        };
    }
    static fromJSON(obj) {
        return new FunctionRequestTriggerParams({
            bounty: (obj.bounty && new common_1.BN(obj.bounty)) || null,
            slotsUntilExpiration: (obj.slotsUntilExpiration && new common_1.BN(obj.slotsUntilExpiration)) || null,
            validAfterSlot: (obj.validAfterSlot && new common_1.BN(obj.validAfterSlot)) || null,
        });
    }
    toEncodable() {
        return FunctionRequestTriggerParams.toEncodable(this);
    }
}
exports.FunctionRequestTriggerParams = FunctionRequestTriggerParams;
