// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionVerifyParams.js
import * as borsh from "@coral-xyz/borsh";
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class FunctionVerifyParams {
    observedTime;
    nextAllowedTimestamp;
    errorCode;
    mrEnclave;
    constructor(fields) {
        this.observedTime = fields.observedTime;
        this.nextAllowedTimestamp = fields.nextAllowedTimestamp;
        this.errorCode = fields.errorCode;
        this.mrEnclave = fields.mrEnclave;
    }
    static layout(property) {
        return borsh.struct([
            borsh.i64("observedTime"),
            borsh.i64("nextAllowedTimestamp"),
            borsh.u8("errorCode"),
            borsh.array(borsh.u8(), 32, "mrEnclave"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionVerifyParams({
            observedTime: obj.observedTime,
            nextAllowedTimestamp: obj.nextAllowedTimestamp,
            errorCode: obj.errorCode,
            mrEnclave: obj.mrEnclave,
        });
    }
    static toEncodable(fields) {
        return {
            observedTime: fields.observedTime,
            nextAllowedTimestamp: fields.nextAllowedTimestamp,
            errorCode: fields.errorCode,
            mrEnclave: fields.mrEnclave,
        };
    }
    toJSON() {
        return {
            observedTime: this.observedTime.toString(),
            nextAllowedTimestamp: this.nextAllowedTimestamp.toString(),
            errorCode: this.errorCode,
            mrEnclave: this.mrEnclave,
        };
    }
    static fromJSON(obj) {
        return new FunctionVerifyParams({
            observedTime: new BN(obj.observedTime),
            nextAllowedTimestamp: new BN(obj.nextAllowedTimestamp),
            errorCode: obj.errorCode,
            mrEnclave: obj.mrEnclave,
        });
    }
    toEncodable() {
        return FunctionVerifyParams.toEncodable(this);
    }
}
