// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionRequestSetConfigParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionRequestSetConfigParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
class FunctionRequestSetConfigParams {
    containerParams;
    appendContainerParams;
    constructor(fields) {
        this.containerParams = fields.containerParams;
        this.appendContainerParams = fields.appendContainerParams;
    }
    static layout(property) {
        return borsh.struct([borsh.vecU8("containerParams"), borsh.bool("appendContainerParams")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionRequestSetConfigParams({
            containerParams: new Uint8Array(obj.containerParams.buffer, obj.containerParams.byteOffset, obj.containerParams.length),
            appendContainerParams: obj.appendContainerParams,
        });
    }
    static toEncodable(fields) {
        return {
            containerParams: Buffer.from(fields.containerParams.buffer, fields.containerParams.byteOffset, fields.containerParams.length),
            appendContainerParams: fields.appendContainerParams,
        };
    }
    toJSON() {
        return {
            containerParams: Array.from(this.containerParams.values()),
            appendContainerParams: this.appendContainerParams,
        };
    }
    static fromJSON(obj) {
        return new FunctionRequestSetConfigParams({
            containerParams: Uint8Array.from(obj.containerParams),
            appendContainerParams: obj.appendContainerParams,
        });
    }
    toEncodable() {
        return FunctionRequestSetConfigParams.toEncodable(this);
    }
}
exports.FunctionRequestSetConfigParams = FunctionRequestSetConfigParams;
