// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionRequestInitParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionRequestInitParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class FunctionRequestInitParams {
    maxContainerParamsLen;
    containerParams;
    garbageCollectionSlot;
    constructor(fields) {
        this.maxContainerParamsLen = fields.maxContainerParamsLen;
        this.containerParams = fields.containerParams;
        this.garbageCollectionSlot = fields.garbageCollectionSlot;
    }
    static layout(property) {
        return borsh.struct([
            borsh.option(borsh.u32(), "maxContainerParamsLen"),
            borsh.vecU8("containerParams"),
            borsh.option(borsh.u64(), "garbageCollectionSlot"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionRequestInitParams({
            maxContainerParamsLen: obj.maxContainerParamsLen,
            containerParams: new Uint8Array(obj.containerParams.buffer, obj.containerParams.byteOffset, obj.containerParams.length),
            garbageCollectionSlot: obj.garbageCollectionSlot,
        });
    }
    static toEncodable(fields) {
        return {
            maxContainerParamsLen: fields.maxContainerParamsLen,
            containerParams: Buffer.from(fields.containerParams.buffer, fields.containerParams.byteOffset, fields.containerParams.length),
            garbageCollectionSlot: fields.garbageCollectionSlot,
        };
    }
    toJSON() {
        return {
            maxContainerParamsLen: this.maxContainerParamsLen,
            containerParams: Array.from(this.containerParams.values()),
            garbageCollectionSlot: (this.garbageCollectionSlot && this.garbageCollectionSlot.toString()) ||
                null,
        };
    }
    static fromJSON(obj) {
        return new FunctionRequestInitParams({
            maxContainerParamsLen: obj.maxContainerParamsLen,
            containerParams: Uint8Array.from(obj.containerParams),
            garbageCollectionSlot: (obj.garbageCollectionSlot && new common_1.BN(obj.garbageCollectionSlot)) ||
                null,
        });
    }
    toEncodable() {
        return FunctionRequestInitParams.toEncodable(this);
    }
}
exports.FunctionRequestInitParams = FunctionRequestInitParams;
