// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionRequestVerifyParams.js
import * as borsh from "@coral-xyz/borsh";
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class FunctionRequestVerifyParams {
    observedTime;
    errorCode;
    mrEnclave;
    requestSlot;
    containerParamsHash;
    constructor(fields) {
        this.observedTime = fields.observedTime;
        this.errorCode = fields.errorCode;
        this.mrEnclave = fields.mrEnclave;
        this.requestSlot = fields.requestSlot;
        this.containerParamsHash = fields.containerParamsHash;
    }
    static layout(property) {
        return borsh.struct([
            borsh.i64("observedTime"),
            borsh.u8("errorCode"),
            borsh.array(borsh.u8(), 32, "mrEnclave"),
            borsh.u64("requestSlot"),
            borsh.array(borsh.u8(), 32, "containerParamsHash"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionRequestVerifyParams({
            observedTime: obj.observedTime,
            errorCode: obj.errorCode,
            mrEnclave: obj.mrEnclave,
            requestSlot: obj.requestSlot,
            containerParamsHash: obj.containerParamsHash,
        });
    }
    static toEncodable(fields) {
        return {
            observedTime: fields.observedTime,
            errorCode: fields.errorCode,
            mrEnclave: fields.mrEnclave,
            requestSlot: fields.requestSlot,
            containerParamsHash: fields.containerParamsHash,
        };
    }
    toJSON() {
        return {
            observedTime: this.observedTime.toString(),
            errorCode: this.errorCode,
            mrEnclave: this.mrEnclave,
            requestSlot: this.requestSlot.toString(),
            containerParamsHash: this.containerParamsHash,
        };
    }
    static fromJSON(obj) {
        return new FunctionRequestVerifyParams({
            observedTime: new BN(obj.observedTime),
            errorCode: obj.errorCode,
            mrEnclave: obj.mrEnclave,
            requestSlot: new BN(obj.requestSlot),
            containerParamsHash: obj.containerParamsHash,
        });
    }
    toEncodable() {
        return FunctionRequestVerifyParams.toEncodable(this);
    }
}
