// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionRequestInitParams.js
import * as borsh from "@coral-xyz/borsh";
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class FunctionRequestInitParams {
    maxContainerParamsLen;
    containerParams;
    garbageCollectionSlot;
    constructor(fields) {
        this.maxContainerParamsLen = fields.maxContainerParamsLen;
        this.containerParams = fields.containerParams;
        this.garbageCollectionSlot = fields.garbageCollectionSlot;
    }
    static layout(property) {
        return borsh.struct([
            borsh.option(borsh.u32(), "maxContainerParamsLen"),
            borsh.vecU8("containerParams"),
            borsh.option(borsh.u64(), "garbageCollectionSlot"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionRequestInitParams({
            maxContainerParamsLen: obj.maxContainerParamsLen,
            containerParams: new Uint8Array(obj.containerParams.buffer, obj.containerParams.byteOffset, obj.containerParams.length),
            garbageCollectionSlot: obj.garbageCollectionSlot,
        });
    }
    static toEncodable(fields) {
        return {
            maxContainerParamsLen: fields.maxContainerParamsLen,
            containerParams: Buffer.from(fields.containerParams.buffer, fields.containerParams.byteOffset, fields.containerParams.length),
            garbageCollectionSlot: fields.garbageCollectionSlot,
        };
    }
    toJSON() {
        return {
            maxContainerParamsLen: this.maxContainerParamsLen,
            containerParams: Array.from(this.containerParams.values()),
            garbageCollectionSlot: (this.garbageCollectionSlot && this.garbageCollectionSlot.toString()) ||
                null,
        };
    }
    static fromJSON(obj) {
        return new FunctionRequestInitParams({
            maxContainerParamsLen: obj.maxContainerParamsLen,
            containerParams: Uint8Array.from(obj.containerParams),
            garbageCollectionSlot: (obj.garbageCollectionSlot && new BN(obj.garbageCollectionSlot)) ||
                null,
        });
    }
    toEncodable() {
        return FunctionRequestInitParams.toEncodable(this);
    }
}
