// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionExtendLookupParams.js
import * as borsh from "@coral-xyz/borsh";
import { PublicKey } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class FunctionExtendLookupParams {
    newAddresses;
    constructor(fields) {
        this.newAddresses = fields.newAddresses;
    }
    static layout(property) {
        return borsh.struct([borsh.vec(borsh.publicKey(), "newAddresses")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionExtendLookupParams({
            newAddresses: obj.newAddresses,
        });
    }
    static toEncodable(fields) {
        return {
            newAddresses: fields.newAddresses,
        };
    }
    toJSON() {
        return {
            newAddresses: this.newAddresses.map((item) => item.toString()),
        };
    }
    static fromJSON(obj) {
        return new FunctionExtendLookupParams({
            newAddresses: obj.newAddresses.map((item) => new PublicKey(item)),
        });
    }
    toEncodable() {
        return FunctionExtendLookupParams.toEncodable(this);
    }
}
