// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionServiceVerifyParams.js
import * as borsh from "@coral-xyz/borsh";
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class FunctionServiceVerifyParams {
    observedTime;
    mrEnclave;
    constructor(fields) {
        this.observedTime = fields.observedTime;
        this.mrEnclave = fields.mrEnclave;
    }
    static layout(property) {
        return borsh.struct([borsh.i64("observedTime"), borsh.array(borsh.u8(), 32, "mrEnclave")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionServiceVerifyParams({
            observedTime: obj.observedTime,
            mrEnclave: obj.mrEnclave,
        });
    }
    static toEncodable(fields) {
        return {
            observedTime: fields.observedTime,
            mrEnclave: fields.mrEnclave,
        };
    }
    toJSON() {
        return {
            observedTime: this.observedTime.toString(),
            mrEnclave: this.mrEnclave,
        };
    }
    static fromJSON(obj) {
        return new FunctionServiceVerifyParams({
            observedTime: new BN(obj.observedTime),
            mrEnclave: obj.mrEnclave,
        });
    }
    toEncodable() {
        return FunctionServiceVerifyParams.toEncodable(this);
    }
}
