// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/FunctionServiceVerifyParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionServiceVerifyParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class FunctionServiceVerifyParams {
    observedTime;
    mrEnclave;
    constructor(fields) {
        this.observedTime = fields.observedTime;
        this.mrEnclave = fields.mrEnclave;
    }
    static layout(property) {
        return borsh.struct([borsh.i64("observedTime"), borsh.array(borsh.u8(), 32, "mrEnclave")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new FunctionServiceVerifyParams({
            observedTime: obj.observedTime,
            mrEnclave: obj.mrEnclave,
        });
    }
    static toEncodable(fields) {
        return {
            observedTime: fields.observedTime,
            mrEnclave: fields.mrEnclave,
        };
    }
    toJSON() {
        return {
            observedTime: this.observedTime.toString(),
            mrEnclave: this.mrEnclave,
        };
    }
    static fromJSON(obj) {
        return new FunctionServiceVerifyParams({
            observedTime: new common_1.BN(obj.observedTime),
            mrEnclave: obj.mrEnclave,
        });
    }
    toEncodable() {
        return FunctionServiceVerifyParams.toEncodable(this);
    }
}
exports.FunctionServiceVerifyParams = FunctionServiceVerifyParams;
