// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/Quote.js
import * as borsh from "@coral-xyz/borsh";
import { PublicKey } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class Quote {
    /** The address of the signer generated within an enclave. */
    enclaveSigner;
    /** The quotes MRENCLAVE measurement dictating the contents of the secure enclave. */
    mrEnclave;
    /** The VerificationStatus of the quote. */
    verificationStatus;
    /** The unix timestamp when the quote was last verified. */
    verificationTimestamp;
    /** The unix timestamp when the quotes verification status expires. */
    validUntil;
    /** The off-chain registry where the verifiers quote can be located. */
    quoteRegistry;
    /** Key to lookup the buffer data on IPFS or an alternative decentralized storage solution. */
    registryKey;
    /** Reserved. */
    ebuf;
    constructor(fields) {
        this.enclaveSigner = fields.enclaveSigner;
        this.mrEnclave = fields.mrEnclave;
        this.verificationStatus = fields.verificationStatus;
        this.verificationTimestamp = fields.verificationTimestamp;
        this.validUntil = fields.validUntil;
        this.quoteRegistry = fields.quoteRegistry;
        this.registryKey = fields.registryKey;
        this.ebuf = fields.ebuf;
    }
    static layout(property) {
        return borsh.struct([
            borsh.publicKey("enclaveSigner"),
            borsh.array(borsh.u8(), 32, "mrEnclave"),
            borsh.u8("verificationStatus"),
            borsh.i64("verificationTimestamp"),
            borsh.i64("validUntil"),
            borsh.array(borsh.u8(), 32, "quoteRegistry"),
            borsh.array(borsh.u8(), 64, "registryKey"),
            borsh.array(borsh.u8(), 256, "ebuf"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new Quote({
            enclaveSigner: obj.enclaveSigner,
            mrEnclave: obj.mrEnclave,
            verificationStatus: obj.verificationStatus,
            verificationTimestamp: obj.verificationTimestamp,
            validUntil: obj.validUntil,
            quoteRegistry: obj.quoteRegistry,
            registryKey: obj.registryKey,
            ebuf: obj.ebuf,
        });
    }
    static toEncodable(fields) {
        return {
            enclaveSigner: fields.enclaveSigner,
            mrEnclave: fields.mrEnclave,
            verificationStatus: fields.verificationStatus,
            verificationTimestamp: fields.verificationTimestamp,
            validUntil: fields.validUntil,
            quoteRegistry: fields.quoteRegistry,
            registryKey: fields.registryKey,
            ebuf: fields.ebuf,
        };
    }
    toJSON() {
        return {
            enclaveSigner: this.enclaveSigner.toString(),
            mrEnclave: this.mrEnclave,
            verificationStatus: this.verificationStatus,
            verificationTimestamp: this.verificationTimestamp.toString(),
            validUntil: this.validUntil.toString(),
            quoteRegistry: this.quoteRegistry,
            registryKey: this.registryKey,
            ebuf: this.ebuf,
        };
    }
    static fromJSON(obj) {
        return new Quote({
            enclaveSigner: new PublicKey(obj.enclaveSigner),
            mrEnclave: obj.mrEnclave,
            verificationStatus: obj.verificationStatus,
            verificationTimestamp: new BN(obj.verificationTimestamp),
            validUntil: new BN(obj.validUntil),
            quoteRegistry: obj.quoteRegistry,
            registryKey: obj.registryKey,
            ebuf: obj.ebuf,
        });
    }
    toEncodable() {
        return Quote.toEncodable(this);
    }
}
