// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/RequestStatus.js
import * as borsh from "@coral-xyz/borsh";
export class None {
    static discriminator = 0;
    static kind = "None";
    discriminator = 0;
    kind = "None";
    toJSON() {
        return {
            kind: "None",
        };
    }
    toEncodable() {
        return {
            None: {},
        };
    }
}
export class RequestPending {
    static discriminator = 1;
    static kind = "RequestPending";
    discriminator = 1;
    kind = "RequestPending";
    toJSON() {
        return {
            kind: "RequestPending",
        };
    }
    toEncodable() {
        return {
            RequestPending: {},
        };
    }
}
export class RequestCancelled {
    static discriminator = 2;
    static kind = "RequestCancelled";
    discriminator = 2;
    kind = "RequestCancelled";
    toJSON() {
        return {
            kind: "RequestCancelled",
        };
    }
    toEncodable() {
        return {
            RequestCancelled: {},
        };
    }
}
export class RequestFailure {
    static discriminator = 3;
    static kind = "RequestFailure";
    discriminator = 3;
    kind = "RequestFailure";
    toJSON() {
        return {
            kind: "RequestFailure",
        };
    }
    toEncodable() {
        return {
            RequestFailure: {},
        };
    }
}
export class RequestExpired {
    static discriminator = 4;
    static kind = "RequestExpired";
    discriminator = 4;
    kind = "RequestExpired";
    toJSON() {
        return {
            kind: "RequestExpired",
        };
    }
    toEncodable() {
        return {
            RequestExpired: {},
        };
    }
}
export class RequestSuccess {
    static discriminator = 5;
    static kind = "RequestSuccess";
    discriminator = 5;
    kind = "RequestSuccess";
    toJSON() {
        return {
            kind: "RequestSuccess",
        };
    }
    toEncodable() {
        return {
            RequestSuccess: {},
        };
    }
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("None" in obj) {
        return new None();
    }
    if ("RequestPending" in obj) {
        return new RequestPending();
    }
    if ("RequestCancelled" in obj) {
        return new RequestCancelled();
    }
    if ("RequestFailure" in obj) {
        return new RequestFailure();
    }
    if ("RequestExpired" in obj) {
        return new RequestExpired();
    }
    if ("RequestSuccess" in obj) {
        return new RequestSuccess();
    }
    throw new Error("Invalid enum object");
}
export function fromJSON(obj) {
    switch (obj.kind) {
        case "None": {
            return new None();
        }
        case "RequestPending": {
            return new RequestPending();
        }
        case "RequestCancelled": {
            return new RequestCancelled();
        }
        case "RequestFailure": {
            return new RequestFailure();
        }
        case "RequestExpired": {
            return new RequestExpired();
        }
        case "RequestSuccess": {
            return new RequestSuccess();
        }
    }
}
export function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "None"),
        borsh.struct([], "RequestPending"),
        borsh.struct([], "RequestCancelled"),
        borsh.struct([], "RequestFailure"),
        borsh.struct([], "RequestExpired"),
        borsh.struct([], "RequestSuccess"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
