// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/ResourceLevel.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.layout = exports.fromJSON = exports.fromDecoded = exports.Queue = exports.Function = exports.Authority = exports.None = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
class None {
    static discriminator = 0;
    static kind = "None";
    discriminator = 0;
    kind = "None";
    toJSON() {
        return {
            kind: "None",
        };
    }
    toEncodable() {
        return {
            None: {},
        };
    }
}
exports.None = None;
class Authority {
    static discriminator = 1;
    static kind = "Authority";
    discriminator = 1;
    kind = "Authority";
    toJSON() {
        return {
            kind: "Authority",
        };
    }
    toEncodable() {
        return {
            Authority: {},
        };
    }
}
exports.Authority = Authority;
class Function {
    static discriminator = 2;
    static kind = "Function";
    discriminator = 2;
    kind = "Function";
    toJSON() {
        return {
            kind: "Function",
        };
    }
    toEncodable() {
        return {
            Function: {},
        };
    }
}
exports.Function = Function;
class Queue {
    static discriminator = 3;
    static kind = "Queue";
    discriminator = 3;
    kind = "Queue";
    toJSON() {
        return {
            kind: "Queue",
        };
    }
    toEncodable() {
        return {
            Queue: {},
        };
    }
}
exports.Queue = Queue;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("None" in obj) {
        return new None();
    }
    if ("Authority" in obj) {
        return new Authority();
    }
    if ("Function" in obj) {
        return new Function();
    }
    if ("Queue" in obj) {
        return new Queue();
    }
    throw new Error("Invalid enum object");
}
exports.fromDecoded = fromDecoded;
function fromJSON(obj) {
    switch (obj.kind) {
        case "None": {
            return new None();
        }
        case "Authority": {
            return new Authority();
        }
        case "Function": {
            return new Function();
        }
        case "Queue": {
            return new Queue();
        }
    }
}
exports.fromJSON = fromJSON;
function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "None"),
        borsh.struct([], "Authority"),
        borsh.struct([], "Function"),
        borsh.struct([], "Queue"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
exports.layout = layout;
