// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/RoutineStatus.js
import * as borsh from "@coral-xyz/borsh";
export class None {
    static discriminator = 0;
    static kind = "None";
    discriminator = 0;
    kind = "None";
    toJSON() {
        return {
            kind: "None",
        };
    }
    toEncodable() {
        return {
            None: {},
        };
    }
}
export class Active {
    static discriminator = 1;
    static kind = "Active";
    discriminator = 1;
    kind = "Active";
    toJSON() {
        return {
            kind: "Active",
        };
    }
    toEncodable() {
        return {
            Active: {},
        };
    }
}
export class NonExecutable {
    static discriminator = 2;
    static kind = "NonExecutable";
    discriminator = 2;
    kind = "NonExecutable";
    toJSON() {
        return {
            kind: "NonExecutable",
        };
    }
    toEncodable() {
        return {
            NonExecutable: {},
        };
    }
}
export class OutOfFunds {
    static discriminator = 3;
    static kind = "OutOfFunds";
    discriminator = 3;
    kind = "OutOfFunds";
    toJSON() {
        return {
            kind: "OutOfFunds",
        };
    }
    toEncodable() {
        return {
            OutOfFunds: {},
        };
    }
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("None" in obj) {
        return new None();
    }
    if ("Active" in obj) {
        return new Active();
    }
    if ("NonExecutable" in obj) {
        return new NonExecutable();
    }
    if ("OutOfFunds" in obj) {
        return new OutOfFunds();
    }
    throw new Error("Invalid enum object");
}
export function fromJSON(obj) {
    switch (obj.kind) {
        case "None": {
            return new None();
        }
        case "Active": {
            return new Active();
        }
        case "NonExecutable": {
            return new NonExecutable();
        }
        case "OutOfFunds": {
            return new OutOfFunds();
        }
    }
}
export function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "None"),
        borsh.struct([], "Active"),
        borsh.struct([], "NonExecutable"),
        borsh.struct([], "OutOfFunds"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
