// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/ServiceRow.js
import * as borsh from "@coral-xyz/borsh";
import { PublicKey } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class ServiceRow {
    service;
    function;
    constructor(fields) {
        this.service = fields.service;
        this.function = fields.function;
    }
    static layout(property) {
        return borsh.struct([borsh.publicKey("service"), borsh.publicKey("function")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new ServiceRow({
            service: obj.service,
            function: obj.function,
        });
    }
    static toEncodable(fields) {
        return {
            service: fields.service,
            function: fields.function,
        };
    }
    toJSON() {
        return {
            service: this.service.toString(),
            function: this.function.toString(),
        };
    }
    static fromJSON(obj) {
        return new ServiceRow({
            service: new PublicKey(obj.service),
            function: new PublicKey(obj.function),
        });
    }
    toEncodable() {
        return ServiceRow.toEncodable(this);
    }
}
