// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/ServiceRequestQuoteVerifyParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceRequestQuoteVerifyParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
class ServiceRequestQuoteVerifyParams {
    quoteRegistry;
    registryKey;
    constructor(fields) {
        this.quoteRegistry = fields.quoteRegistry;
        this.registryKey = fields.registryKey;
    }
    static layout(property) {
        return borsh.struct([
            borsh.option(borsh.vecU8(), "quoteRegistry"),
            borsh.vecU8("registryKey"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new ServiceRequestQuoteVerifyParams({
            quoteRegistry: (obj.quoteRegistry &&
                new Uint8Array(obj.quoteRegistry.buffer, obj.quoteRegistry.byteOffset, obj.quoteRegistry.length)) ||
                null,
            registryKey: new Uint8Array(obj.registryKey.buffer, obj.registryKey.byteOffset, obj.registryKey.length),
        });
    }
    static toEncodable(fields) {
        return {
            quoteRegistry: (fields.quoteRegistry &&
                Buffer.from(fields.quoteRegistry.buffer, fields.quoteRegistry.byteOffset, fields.quoteRegistry.length)) ||
                null,
            registryKey: Buffer.from(fields.registryKey.buffer, fields.registryKey.byteOffset, fields.registryKey.length),
        };
    }
    toJSON() {
        return {
            quoteRegistry: (this.quoteRegistry && Array.from(this.quoteRegistry.values())) || null,
            registryKey: Array.from(this.registryKey.values()),
        };
    }
    static fromJSON(obj) {
        return new ServiceRequestQuoteVerifyParams({
            quoteRegistry: (obj.quoteRegistry && Uint8Array.from(obj.quoteRegistry)) || null,
            registryKey: Uint8Array.from(obj.registryKey),
        });
    }
    toEncodable() {
        return ServiceRequestQuoteVerifyParams.toEncodable(this);
    }
}
exports.ServiceRequestQuoteVerifyParams = ServiceRequestQuoteVerifyParams;
