// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/ServiceRequestQuoteVerifyParams.js
import * as borsh from "@coral-xyz/borsh";
export class ServiceRequestQuoteVerifyParams {
    quoteRegistry;
    registryKey;
    constructor(fields) {
        this.quoteRegistry = fields.quoteRegistry;
        this.registryKey = fields.registryKey;
    }
    static layout(property) {
        return borsh.struct([
            borsh.option(borsh.vecU8(), "quoteRegistry"),
            borsh.vecU8("registryKey"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new ServiceRequestQuoteVerifyParams({
            quoteRegistry: (obj.quoteRegistry &&
                new Uint8Array(obj.quoteRegistry.buffer, obj.quoteRegistry.byteOffset, obj.quoteRegistry.length)) ||
                null,
            registryKey: new Uint8Array(obj.registryKey.buffer, obj.registryKey.byteOffset, obj.registryKey.length),
        });
    }
    static toEncodable(fields) {
        return {
            quoteRegistry: (fields.quoteRegistry &&
                Buffer.from(fields.quoteRegistry.buffer, fields.quoteRegistry.byteOffset, fields.quoteRegistry.length)) ||
                null,
            registryKey: Buffer.from(fields.registryKey.buffer, fields.registryKey.byteOffset, fields.registryKey.length),
        };
    }
    toJSON() {
        return {
            quoteRegistry: (this.quoteRegistry && Array.from(this.quoteRegistry.values())) || null,
            registryKey: Array.from(this.registryKey.values()),
        };
    }
    static fromJSON(obj) {
        return new ServiceRequestQuoteVerifyParams({
            quoteRegistry: (obj.quoteRegistry && Uint8Array.from(obj.quoteRegistry)) || null,
            registryKey: Uint8Array.from(obj.registryKey),
        });
    }
    toEncodable() {
        return ServiceRequestQuoteVerifyParams.toEncodable(this);
    }
}
