// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/ServiceWorkerSetConfigParams.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh";
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class ServiceWorkerSetConfigParams {
    region;
    zone;
    permissionsRequired;
    lockPermissionsRequired;
    availableEnclaveSize;
    maxEnclaveSize;
    maxCpu;
    enclaveCost;
    constructor(fields) {
        this.region = fields.region;
        this.zone = fields.zone;
        this.permissionsRequired = fields.permissionsRequired;
        this.lockPermissionsRequired = fields.lockPermissionsRequired;
        this.availableEnclaveSize = fields.availableEnclaveSize;
        this.maxEnclaveSize = fields.maxEnclaveSize;
        this.maxCpu = fields.maxCpu;
        this.enclaveCost = fields.enclaveCost;
    }
    static layout(property) {
        return borsh.struct([
            borsh.option(types.ServerRegion.layout(), "region"),
            borsh.option(types.ServerZone.layout(), "zone"),
            borsh.option(borsh.bool(), "permissionsRequired"),
            borsh.option(borsh.bool(), "lockPermissionsRequired"),
            borsh.option(borsh.u64(), "availableEnclaveSize"),
            borsh.option(borsh.u64(), "maxEnclaveSize"),
            borsh.option(borsh.u64(), "maxCpu"),
            borsh.option(borsh.u64(), "enclaveCost"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new ServiceWorkerSetConfigParams({
            region: (obj.region && types.ServerRegion.fromDecoded(obj.region)) || null,
            zone: (obj.zone && types.ServerZone.fromDecoded(obj.zone)) || null,
            permissionsRequired: obj.permissionsRequired,
            lockPermissionsRequired: obj.lockPermissionsRequired,
            availableEnclaveSize: obj.availableEnclaveSize,
            maxEnclaveSize: obj.maxEnclaveSize,
            maxCpu: obj.maxCpu,
            enclaveCost: obj.enclaveCost,
        });
    }
    static toEncodable(fields) {
        return {
            region: (fields.region && fields.region.toEncodable()) || null,
            zone: (fields.zone && fields.zone.toEncodable()) || null,
            permissionsRequired: fields.permissionsRequired,
            lockPermissionsRequired: fields.lockPermissionsRequired,
            availableEnclaveSize: fields.availableEnclaveSize,
            maxEnclaveSize: fields.maxEnclaveSize,
            maxCpu: fields.maxCpu,
            enclaveCost: fields.enclaveCost,
        };
    }
    toJSON() {
        return {
            region: (this.region && this.region.toJSON()) || null,
            zone: (this.zone && this.zone.toJSON()) || null,
            permissionsRequired: this.permissionsRequired,
            lockPermissionsRequired: this.lockPermissionsRequired,
            availableEnclaveSize: (this.availableEnclaveSize && this.availableEnclaveSize.toString()) ||
                null,
            maxEnclaveSize: (this.maxEnclaveSize && this.maxEnclaveSize.toString()) || null,
            maxCpu: (this.maxCpu && this.maxCpu.toString()) || null,
            enclaveCost: (this.enclaveCost && this.enclaveCost.toString()) || null,
        };
    }
    static fromJSON(obj) {
        return new ServiceWorkerSetConfigParams({
            region: (obj.region && types.ServerRegion.fromJSON(obj.region)) || null,
            zone: (obj.zone && types.ServerZone.fromJSON(obj.zone)) || null,
            permissionsRequired: obj.permissionsRequired,
            lockPermissionsRequired: obj.lockPermissionsRequired,
            availableEnclaveSize: (obj.availableEnclaveSize && new BN(obj.availableEnclaveSize)) || null,
            maxEnclaveSize: (obj.maxEnclaveSize && new BN(obj.maxEnclaveSize)) || null,
            maxCpu: (obj.maxCpu && new BN(obj.maxCpu)) || null,
            enclaveCost: (obj.enclaveCost && new BN(obj.enclaveCost)) || null,
        });
    }
    toEncodable() {
        return ServiceWorkerSetConfigParams.toEncodable(this);
    }
}
