// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/ServiceRow.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceRow = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
const web3_js_1 = require("@solana/web3.js"); // eslint-disable-line @typescript-eslint/no-unused-vars
class ServiceRow {
    service;
    function;
    constructor(fields) {
        this.service = fields.service;
        this.function = fields.function;
    }
    static layout(property) {
        return borsh.struct([borsh.publicKey("service"), borsh.publicKey("function")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new ServiceRow({
            service: obj.service,
            function: obj.function,
        });
    }
    static toEncodable(fields) {
        return {
            service: fields.service,
            function: fields.function,
        };
    }
    toJSON() {
        return {
            service: this.service.toString(),
            function: this.function.toString(),
        };
    }
    static fromJSON(obj) {
        return new ServiceRow({
            service: new web3_js_1.PublicKey(obj.service),
            function: new web3_js_1.PublicKey(obj.function),
        });
    }
    toEncodable() {
        return ServiceRow.toEncodable(this);
    }
}
exports.ServiceRow = ServiceRow;
