// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/SwitchboardAttestationPermission.js
import * as borsh from "@coral-xyz/borsh";
export class None {
    static discriminator = 0;
    static kind = "None";
    discriminator = 0;
    kind = "None";
    toJSON() {
        return {
            kind: "None",
        };
    }
    toEncodable() {
        return {
            None: {},
        };
    }
}
export class PermitNodeheartbeat {
    static discriminator = 1;
    static kind = "PermitNodeheartbeat";
    discriminator = 1;
    kind = "PermitNodeheartbeat";
    toJSON() {
        return {
            kind: "PermitNodeheartbeat",
        };
    }
    toEncodable() {
        return {
            PermitNodeheartbeat: {},
        };
    }
}
export class PermitQueueUsage {
    static discriminator = 2;
    static kind = "PermitQueueUsage";
    discriminator = 2;
    kind = "PermitQueueUsage";
    toJSON() {
        return {
            kind: "PermitQueueUsage",
        };
    }
    toEncodable() {
        return {
            PermitQueueUsage: {},
        };
    }
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function fromDecoded(obj) {
    if (typeof obj !== "object") {
        throw new Error("Invalid enum object");
    }
    if ("None" in obj) {
        return new None();
    }
    if ("PermitNodeheartbeat" in obj) {
        return new PermitNodeheartbeat();
    }
    if ("PermitQueueUsage" in obj) {
        return new PermitQueueUsage();
    }
    throw new Error("Invalid enum object");
}
export function fromJSON(obj) {
    switch (obj.kind) {
        case "None": {
            return new None();
        }
        case "PermitNodeheartbeat": {
            return new PermitNodeheartbeat();
        }
        case "PermitQueueUsage": {
            return new PermitQueueUsage();
        }
    }
}
export function layout(property) {
    const ret = borsh.rustEnum([
        borsh.struct([], "None"),
        borsh.struct([], "PermitNodeheartbeat"),
        borsh.struct([], "PermitQueueUsage"),
    ]);
    if (property !== undefined) {
        return ret.replicate(property);
    }
    return ret;
}
