// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/VerifierQuoteVerifyParams.js
import * as borsh from "@coral-xyz/borsh";
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class VerifierQuoteVerifyParams {
    timestamp;
    mrEnclave;
    idx;
    constructor(fields) {
        this.timestamp = fields.timestamp;
        this.mrEnclave = fields.mrEnclave;
        this.idx = fields.idx;
    }
    static layout(property) {
        return borsh.struct([
            borsh.i64("timestamp"),
            borsh.array(borsh.u8(), 32, "mrEnclave"),
            borsh.u32("idx"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new VerifierQuoteVerifyParams({
            timestamp: obj.timestamp,
            mrEnclave: obj.mrEnclave,
            idx: obj.idx,
        });
    }
    static toEncodable(fields) {
        return {
            timestamp: fields.timestamp,
            mrEnclave: fields.mrEnclave,
            idx: fields.idx,
        };
    }
    toJSON() {
        return {
            timestamp: this.timestamp.toString(),
            mrEnclave: this.mrEnclave,
            idx: this.idx,
        };
    }
    static fromJSON(obj) {
        return new VerifierQuoteVerifyParams({
            timestamp: new BN(obj.timestamp),
            mrEnclave: obj.mrEnclave,
            idx: obj.idx,
        });
    }
    toEncodable() {
        return VerifierQuoteVerifyParams.toEncodable(this);
    }
}
