// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/VerifierQuoteVerifyParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerifierQuoteVerifyParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
const common_1 = require("@switchboard-xyz/common"); // eslint-disable-line @typescript-eslint/no-unused-vars
class VerifierQuoteVerifyParams {
    timestamp;
    mrEnclave;
    idx;
    constructor(fields) {
        this.timestamp = fields.timestamp;
        this.mrEnclave = fields.mrEnclave;
        this.idx = fields.idx;
    }
    static layout(property) {
        return borsh.struct([
            borsh.i64("timestamp"),
            borsh.array(borsh.u8(), 32, "mrEnclave"),
            borsh.u32("idx"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new VerifierQuoteVerifyParams({
            timestamp: obj.timestamp,
            mrEnclave: obj.mrEnclave,
            idx: obj.idx,
        });
    }
    static toEncodable(fields) {
        return {
            timestamp: fields.timestamp,
            mrEnclave: fields.mrEnclave,
            idx: fields.idx,
        };
    }
    toJSON() {
        return {
            timestamp: this.timestamp.toString(),
            mrEnclave: this.mrEnclave,
            idx: this.idx,
        };
    }
    static fromJSON(obj) {
        return new VerifierQuoteVerifyParams({
            timestamp: new common_1.BN(obj.timestamp),
            mrEnclave: obj.mrEnclave,
            idx: obj.idx,
        });
    }
    toEncodable() {
        return VerifierQuoteVerifyParams.toEncodable(this);
    }
}
exports.VerifierQuoteVerifyParams = VerifierQuoteVerifyParams;
