// @switchboard-xyz/solana.js/lib/generated/attestation-program/types/WalletFundParams.js
import * as borsh from "@coral-xyz/borsh";
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class WalletFundParams {
    transferAmount;
    wrapAmount;
    constructor(fields) {
        this.transferAmount = fields.transferAmount;
        this.wrapAmount = fields.wrapAmount;
    }
    static layout(property) {
        return borsh.struct([
            borsh.option(borsh.u64(), "transferAmount"),
            borsh.option(borsh.u64(), "wrapAmount"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new WalletFundParams({
            transferAmount: obj.transferAmount,
            wrapAmount: obj.wrapAmount,
        });
    }
    static toEncodable(fields) {
        return {
            transferAmount: fields.transferAmount,
            wrapAmount: fields.wrapAmount,
        };
    }
    toJSON() {
        return {
            transferAmount: (this.transferAmount && this.transferAmount.toString()) || null,
            wrapAmount: (this.wrapAmount && this.wrapAmount.toString()) || null,
        };
    }
    static fromJSON(obj) {
        return new WalletFundParams({
            transferAmount: (obj.transferAmount && new BN(obj.transferAmount)) || null,
            wrapAmount: (obj.wrapAmount && new BN(obj.wrapAmount)) || null,
        });
    }
    toEncodable() {
        return WalletFundParams.toEncodable(this);
    }
}
