// @switchboard-xyz/solana.js/lib/generated/oracle-program/types/AggregatorFunctionUpsertParams.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh";
export class AggregatorFunctionUpsertParams {
    name;
    value;
    constructor(fields) {
        this.name = fields.name;
        this.value = new types.BorshDecimal({ ...fields.value });
    }
    static layout(property) {
        return borsh.struct([borsh.array(borsh.u8(), 32, "name"), types.BorshDecimal.layout("value")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new AggregatorFunctionUpsertParams({
            name: obj.name,
            value: types.BorshDecimal.fromDecoded(obj.value),
        });
    }
    static toEncodable(fields) {
        return {
            name: fields.name,
            value: types.BorshDecimal.toEncodable(fields.value),
        };
    }
    toJSON() {
        return {
            name: this.name,
            value: this.value.toJSON(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorFunctionUpsertParams({
            name: obj.name,
            value: types.BorshDecimal.fromJSON(obj.value),
        });
    }
    toEncodable() {
        return AggregatorFunctionUpsertParams.toEncodable(this);
    }
}
