// @switchboard-xyz/solana.js/lib/generated/oracle-program/types/AggregatorFunctionUpsertParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AggregatorFunctionUpsertParams = void 0;
const types = __importStar(require("../types/index.cjs")); // eslint-disable-line @typescript-eslint/no-unused-vars
const borsh = __importStar(require("@coral-xyz/borsh"));
class AggregatorFunctionUpsertParams {
    name;
    value;
    constructor(fields) {
        this.name = fields.name;
        this.value = new types.BorshDecimal({ ...fields.value });
    }
    static layout(property) {
        return borsh.struct([borsh.array(borsh.u8(), 32, "name"), types.BorshDecimal.layout("value")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new AggregatorFunctionUpsertParams({
            name: obj.name,
            value: types.BorshDecimal.fromDecoded(obj.value),
        });
    }
    static toEncodable(fields) {
        return {
            name: fields.name,
            value: types.BorshDecimal.toEncodable(fields.value),
        };
    }
    toJSON() {
        return {
            name: this.name,
            value: this.value.toJSON(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorFunctionUpsertParams({
            name: obj.name,
            value: types.BorshDecimal.fromJSON(obj.value),
        });
    }
    toEncodable() {
        return AggregatorFunctionUpsertParams.toEncodable(this);
    }
}
exports.AggregatorFunctionUpsertParams = AggregatorFunctionUpsertParams;
