// @switchboard-xyz/solana.js/lib/generated/oracle-program/types/AggregatorSetVarianceThresholdParams.js
import * as types from "../types/index.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh";
export class AggregatorSetVarianceThresholdParams {
    varianceThreshold;
    constructor(fields) {
        this.varianceThreshold = new types.BorshDecimal({
            ...fields.varianceThreshold,
        });
    }
    static layout(property) {
        return borsh.struct([types.BorshDecimal.layout("varianceThreshold")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new AggregatorSetVarianceThresholdParams({
            varianceThreshold: types.BorshDecimal.fromDecoded(obj.varianceThreshold),
        });
    }
    static toEncodable(fields) {
        return {
            varianceThreshold: types.BorshDecimal.toEncodable(fields.varianceThreshold),
        };
    }
    toJSON() {
        return {
            varianceThreshold: this.varianceThreshold.toJSON(),
        };
    }
    static fromJSON(obj) {
        return new AggregatorSetVarianceThresholdParams({
            varianceThreshold: types.BorshDecimal.fromJSON(obj.varianceThreshold),
        });
    }
    toEncodable() {
        return AggregatorSetVarianceThresholdParams.toEncodable(this);
    }
}
