// @switchboard-xyz/solana.js/lib/generated/oracle-program/types/BorshDecimal.js
import * as borsh from "@coral-xyz/borsh";
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class BorshDecimal {
    mantissa;
    scale;
    constructor(fields) {
        this.mantissa = fields.mantissa;
        this.scale = fields.scale;
    }
    static layout(property) {
        return borsh.struct([borsh.i128("mantissa"), borsh.u32("scale")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new BorshDecimal({
            mantissa: obj.mantissa,
            scale: obj.scale,
        });
    }
    static toEncodable(fields) {
        return {
            mantissa: fields.mantissa,
            scale: fields.scale,
        };
    }
    toJSON() {
        return {
            mantissa: this.mantissa.toString(),
            scale: this.scale,
        };
    }
    static fromJSON(obj) {
        return new BorshDecimal({
            mantissa: new BN(obj.mantissa),
            scale: obj.scale,
        });
    }
    toEncodable() {
        return BorshDecimal.toEncodable(this);
    }
}
