// @switchboard-xyz/solana.js/lib/generated/oracle-program/types/CrankRow.js
import * as borsh from "@coral-xyz/borsh";
import { PublicKey } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
import { BN } from "@switchboard-xyz/common"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class CrankRow {
    /** The PublicKey of the AggregatorAccountData. */
    pubkey;
    /** The aggregator's next available update time. */
    nextTimestamp;
    constructor(fields) {
        this.pubkey = fields.pubkey;
        this.nextTimestamp = fields.nextTimestamp;
    }
    static layout(property) {
        return borsh.struct([borsh.publicKey("pubkey"), borsh.i64("nextTimestamp")], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new CrankRow({
            pubkey: obj.pubkey,
            nextTimestamp: obj.nextTimestamp,
        });
    }
    static toEncodable(fields) {
        return {
            pubkey: fields.pubkey,
            nextTimestamp: fields.nextTimestamp,
        };
    }
    toJSON() {
        return {
            pubkey: this.pubkey.toString(),
            nextTimestamp: this.nextTimestamp.toString(),
        };
    }
    static fromJSON(obj) {
        return new CrankRow({
            pubkey: new PublicKey(obj.pubkey),
            nextTimestamp: new BN(obj.nextTimestamp),
        });
    }
    toEncodable() {
        return CrankRow.toEncodable(this);
    }
}
