// @switchboard-xyz/solana.js/lib/generated/oracle-program/types/ProgramConfigParams.cjs
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgramConfigParams = void 0;
const borsh = __importStar(require("@coral-xyz/borsh"));
const web3_js_1 = require("@solana/web3.js"); // eslint-disable-line @typescript-eslint/no-unused-vars
class ProgramConfigParams {
    token;
    bump;
    daoMint;
    addEnclaves;
    rmEnclaves;
    constructor(fields) {
        this.token = fields.token;
        this.bump = fields.bump;
        this.daoMint = fields.daoMint;
        this.addEnclaves = fields.addEnclaves;
        this.rmEnclaves = fields.rmEnclaves;
    }
    static layout(property) {
        return borsh.struct([
            borsh.publicKey("token"),
            borsh.u8("bump"),
            borsh.publicKey("daoMint"),
            borsh.vec(borsh.array(borsh.u8(), 32), "addEnclaves"),
            borsh.vec(borsh.array(borsh.u8(), 32), "rmEnclaves"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new ProgramConfigParams({
            token: obj.token,
            bump: obj.bump,
            daoMint: obj.daoMint,
            addEnclaves: obj.addEnclaves,
            rmEnclaves: obj.rmEnclaves,
        });
    }
    static toEncodable(fields) {
        return {
            token: fields.token,
            bump: fields.bump,
            daoMint: fields.daoMint,
            addEnclaves: fields.addEnclaves,
            rmEnclaves: fields.rmEnclaves,
        };
    }
    toJSON() {
        return {
            token: this.token.toString(),
            bump: this.bump,
            daoMint: this.daoMint.toString(),
            addEnclaves: this.addEnclaves,
            rmEnclaves: this.rmEnclaves,
        };
    }
    static fromJSON(obj) {
        return new ProgramConfigParams({
            token: new web3_js_1.PublicKey(obj.token),
            bump: obj.bump,
            daoMint: new web3_js_1.PublicKey(obj.daoMint),
            addEnclaves: obj.addEnclaves,
            rmEnclaves: obj.rmEnclaves,
        });
    }
    toEncodable() {
        return ProgramConfigParams.toEncodable(this);
    }
}
exports.ProgramConfigParams = ProgramConfigParams;
