// @switchboard-xyz/solana.js/lib/generated/oracle-program/types/ProgramConfigParams.js
import * as borsh from "@coral-xyz/borsh";
import { PublicKey } from "@solana/web3.js"; // eslint-disable-line @typescript-eslint/no-unused-vars
export class ProgramConfigParams {
    token;
    bump;
    daoMint;
    addEnclaves;
    rmEnclaves;
    constructor(fields) {
        this.token = fields.token;
        this.bump = fields.bump;
        this.daoMint = fields.daoMint;
        this.addEnclaves = fields.addEnclaves;
        this.rmEnclaves = fields.rmEnclaves;
    }
    static layout(property) {
        return borsh.struct([
            borsh.publicKey("token"),
            borsh.u8("bump"),
            borsh.publicKey("daoMint"),
            borsh.vec(borsh.array(borsh.u8(), 32), "addEnclaves"),
            borsh.vec(borsh.array(borsh.u8(), 32), "rmEnclaves"),
        ], property);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static fromDecoded(obj) {
        return new ProgramConfigParams({
            token: obj.token,
            bump: obj.bump,
            daoMint: obj.daoMint,
            addEnclaves: obj.addEnclaves,
            rmEnclaves: obj.rmEnclaves,
        });
    }
    static toEncodable(fields) {
        return {
            token: fields.token,
            bump: fields.bump,
            daoMint: fields.daoMint,
            addEnclaves: fields.addEnclaves,
            rmEnclaves: fields.rmEnclaves,
        };
    }
    toJSON() {
        return {
            token: this.token.toString(),
            bump: this.bump,
            daoMint: this.daoMint.toString(),
            addEnclaves: this.addEnclaves,
            rmEnclaves: this.rmEnclaves,
        };
    }
    static fromJSON(obj) {
        return new ProgramConfigParams({
            token: new PublicKey(obj.token),
            bump: obj.bump,
            daoMint: new PublicKey(obj.daoMint),
            addEnclaves: obj.addEnclaves,
            rmEnclaves: obj.rmEnclaves,
        });
    }
    toEncodable() {
        return ProgramConfigParams.toEncodable(this);
    }
}
